l1-beacon-client: verify blobs using commitment only (not the proof)

geoknee opened the pull request:
We don't need to do blob proof verification when fetching blobs from the beacon API. In fact, we cannot rely on those proofs being available after Fusaka.

Instead, we can verify the blobs by recomputing the versioned hash and comparing against what was found in the L1 block. This requires us to recompute the kzg commitment, which takes 5ms whereas the existing blob proof verification takes only 1.5ms:

```
# BLOB PROOF VERIFY
go test -benchmem -run=^$ -bench ^BenchmarkGoKZGVerifyBlobProof$ github.com/ethereum/go-ethereum/crypto/kzg4844 -v
goos: darwin
goarch: arm64
pkg: github.com/ethereum/go-ethereum/crypto/kzg4844
cpu: Apple M3 Max
BenchmarkGoKZGVerifyBlobProof
BenchmarkGoKZGVerifyBlobProof-14             747           1,568,556 ns/op          399745 B/op      52 allocs/op

# COMMITMENT COMPUTE
➜  op-geth git:(gk/cell-proof-bench) go test -benchmem -run=^$ -bench ^BenchmarkGoKZGBlobToCommitment$ github.com/ethereum/go-ethereum/crypto/kzg4844 -v -benchtime=100x
goos: darwin
goarch: arm64
pkg: github.com/ethereum/go-ethereum/crypto/kzg4844
cpu: Apple M3 Max
BenchmarkGoKZGBlobToCommitment
BenchmarkGoKZGBlobToCommitment-14            100           5,194,772 ns/op          459577 B/op        327 allocs/op
```

Given that this increased overhead is per _blob_, which contains potentially hundreds of L2 blocks which are then either executed or consolidated, the increase is considered negligible. 

⚠️ This PR also reverts #17713 and https://github.com/ethereum-optimism/optimism/pull/17718

geoknee committed:
recompute blob proof if skipBlobVerification

geoknee committed:
use even simpler verification

geoknee committed:
replace verification logic

niran commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -298,9 +298,13 @@ func blobsFromSidecars(blobSidecars []*eth.BlobSidecar, hashes []eth.IndexedBlob
-			// confirm blob data is valid by verifying its proof against the commitment
-			if err := eth.VerifyBlobProof(&sidecar.Blob, kzg4844.Commitment(sidecar.KZGCommitment), kzg4844.Proof(sidecar.KZGProof)); err != nil {
-				return nil, fmt.Errorf("blob at index %d failed verification: %w", i, err)
+			// recompute the commitment from the blob data and check it matches the commitment from the sidecar
+			recomputedCommitment, err := kzg4844.BlobToCommitment(sidecar.Blob.KZGBlob())
+			if err != nil {
+				return nil, fmt.Errorf("cannot compute KZG commitment for blob %d: %w", i, err)
+			}
+			if recomputedCommitment != kzg4844.Commitment(sidecar.KZGCommitment) {
+				return nil, fmt.Errorf("recomputed commitment %s does not match expected commitment %s", recomputedCommitment, kzg4844.Commitment(sidecar.KZGCommitment))
```
This looks right to me, but it makes the invariant we're looking for transitive instead of explicit. It enforces that:

* `ih.Hash == KZGToVersionedHash(sidecar.KZGCommitment)` (L296)
* `sidecar.KZGCommitment == BlobToCommitment(sidecar.Blob)` (L306)

But we don't actually care about the value of `sidecar.KZGCommitment` that the beacon node has given us now that we don't have `sidecar.KZGProof` to use as a shortcut for verifying the blob. We just want to ensure that `ih.Hash == KZGToVersionedHash(BlobToCommitment(sidecar.Blob))`.

geoknee replied:
Totally agree. I've refactored it to not use the "metadata" from the beacon API at all -- it now only uses the blob data. We anticipate that the beacon API will soon be changed to only return that data anyway. [830e96d](https://github.com/ethereum-optimism/optimism/pull/17725/commits/830e96dfdda0d2cfb0e7f7f3d598f79254f1a4bb)

niran commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -298,9 +298,13 @@ func blobsFromSidecars(blobSidecars []*eth.BlobSidecar, hashes []eth.IndexedBlob
-			if err := eth.VerifyBlobProof(&sidecar.Blob, kzg4844.Commitment(sidecar.KZGCommitment), kzg4844.Proof(sidecar.KZGProof)); err != nil {
-				return nil, fmt.Errorf("blob at index %d failed verification: %w", i, err)
+			// recompute the commitment from the blob data and check it matches the commitment from the sidecar
+			recomputedCommitment, err := kzg4844.BlobToCommitment(sidecar.Blob.KZGBlob())
```
We probably want to emit metrics for this. Recomputing the commitments is supposed to be much slower than verifying proofs the way we were doing it before. I'm assuming it's still fast enough that it won't have any affect on the nodes, but collecting timing data will help us be certain.

geoknee replied:
Agreed. I think recomputing the commitments might be pretty slow, and actually regenerating the proof and using that to verify might be more performant.

geoknee replied:
I ran some benchmarks on my laptop, and computing a commitment takes around 5ms where verifying a proof takes 1.5ms.

geoknee replied:
It takes 6ms to (re)-compute a blob proof, so that isn't looking competitive actually.

geoknee replied:
Latest thinking is that this increased overhead is likely negligible compared to the other timescales involved. If we do want to introduce metrics, this can be done in a followup.

sebastianst replied:
I agree, ~5ms (and net +~3.5ms) per blob is negligible compared to all the other things that are happening in either an op-program run or a verified node during derivation. I don't think we need to add metrics given the existing benchmarks.

geoknee committed:
refactor to make verification more direct

This approach does not use any additional data from the beacon endpoint, aside from the blob itself, so anticipates future API changes where this will be the only data provided.

geoknee committed:
tweak comments

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -279,35 +279,42 @@ func (cl *L1BeaconClient) GetBlobs(ctx context.Context, ref eth.L1BlockRef, hash
 
+// verifyBlobProof verifies that the blob data corresponds to the provided commitment.
+// It recomputes the commitment from the blob data and checks it matches the expected commitment hash.
+func verifyBlobProof(blob *eth.Blob, expectedCommitmentHash common.Hash) error {
```
technically, we're verifying the commitment hash, not the proof

```suggestion
func verifyBlobCommitmentHash(blob *eth.Blob, expectedCommitmentHash common.Hash) error {
```

geoknee replied:
[6548dae](https://github.com/ethereum-optimism/optimism/pull/17725/commits/6548dae11e89428948c718c9845065457e327085)

geoknee committed:
rename to verifyBlob

geoknee committed:
tweak test comments

geoknee committed:
Revert "op-service/L1BeaconClient: allow blob proof verification to be skipped via `l1.beacon.skip-blob-verification` flag (#17713)"

This reverts commit 05ba528ae52a48efc2cde8e0e0390cf47ad5f4ad.

geoknee committed:
Revert "Add `skip-blob-verification` flag to op-program and op-challenger  (#17718)"

This reverts commit 3ee883bcfb1b6ac4717ae430a84e92d08359c21f.

sebastianst commented on op-service/sources/l1_beacon_client.go:
```diff
@@ -272,42 +271,45 @@ func (cl *L1BeaconClient) GetBlobs(ctx context.Context, ref eth.L1BlockRef, hash
 
+// verifyBlob verifies that the blob data corresponds to the provided commitment.
+// It recomputes the commitment from the blob data and checks it matches the expected commitment hash.
+func verifyBlob(blob *eth.Blob, expectedCommitmentHash common.Hash) error {
```
Is this function already sufficiently covered by existing tests?

geoknee replied:
I added another test here [0772cd5](https://github.com/ethereum-optimism/optimism/pull/17725/commits/0772cd5caa4456fe2df4f4e32c879b2c460708f6) just for completeness, but yeah I think it is covered by other tests too.

geoknee committed:
add an extra unit test
